Mathlib/Data/Finsupp/Defs.lean

```diff
@@ -0,0 +1,1258 @@
+#align finsupp.support_nonempty_iff Finsupp.support_nonempty_iff
+
+theorem nonzero_iff_exists {f : α →₀ M} : f ≠ 0 ↔ ∃ a : α, f a ≠ 0 := by
+  simp [← Finsupp.support_eq_empty, Finset.eq_empty_iff_forall_not_mem]
+#align finsupp.nonzero_iff_exists Finsupp.nonzero_iff_exists
+
+theorem card_support_eq_zero {f : α →₀ M} : card f.support = 0 ↔ f = 0 := by simp
+#align finsupp.card_support_eq_zero Finsupp.card_support_eq_zero
+
+instance [DecidableEq α] [DecidableEq M] : DecidableEq (α →₀ M) := fun f g =>
+  decidable_of_iff (f.support = g.support ∧ ∀ a ∈ f.support, f a = g a) ext_iff'.symm
+
+theorem finite_support (f : α →₀ M) : Set.Finite (Function.support f) :=
+  f.fun_support_eq.symm ▸ f.support.finite_toSet
+#align finsupp.finite_support Finsupp.finite_support
+
+/- ./././Mathport/Syntax/Translate/Basic.lean:632:2: warning: expanding binder collection (a «expr ∉ » s) -/
+theorem support_subset_iff {s : Set α} {f : α →₀ M} : ↑f.support ⊆ s ↔ ∀ (a) (_ : a ∉ s), f a = 0 :=
+  by
+  simp only [Set.subset_def, mem_coe, mem_support_iff] <;> exact forall_congr' fun a => not_imp_comm
+#align finsupp.support_subset_iff Finsupp.support_subset_iff
+
+/-- Given `finite α`, `equiv_fun_on_finite` is the `equiv` between `α →₀ β` and `α → β`.
+  (All functions on a finite type are finitely supported.) -/
+@[simps]
+def equivFunOnFinite [Finite α] : (α →₀ M) ≃ (α → M)
+    where
+  toFun := (⇑)
+  invFun f := mk (Function.support f).toFinite.toFinset f fun a => Set.Finite.mem_toFinset _
+  left_inv f := ext fun x => rfl
+  right_inv f := rfl
+#align finsupp.equiv_fun_on_finite Finsupp.equivFunOnFinite
+
+@[simp]
+theorem equivFunOnFinite_symm_coe {α} [Finite α] (f : α →₀ M) : equivFunOnFinite.symm f = f :=
+  equivFunOnFinite.symm_apply_apply f
+#align finsupp.equiv_fun_on_finite_symm_coe Finsupp.equivFunOnFinite_symm_coe
+
+/--
+If `α` has a unique term, the type of finitely supported functions `α →₀ β` is equivalent to `β`.
+-/
+@[simps]
+noncomputable def _root_.Equiv.finsuppUnique {ι : Type _} [Unique ι] : (ι →₀ M) ≃ M :=
+  Finsupp.equivFunOnFinite.trans (Equiv.funUnique ι M)
+#align equiv.finsupp_unique Equiv.finsuppUnique
+
+@[ext]
+theorem unique_ext [Unique α] {f g : α →₀ M} (h : f default = g default) : f = g :=
+  ext fun a => by rwa [Unique.eq_default a]
+#align finsupp.unique_ext Finsupp.unique_ext
+
+theorem unique_ext_iff [Unique α] {f g : α →₀ M} : f = g ↔ f default = g default :=
+  ⟨fun h => h ▸ rfl, unique_ext⟩
+#align finsupp.unique_ext_iff Finsupp.unique_ext_iff
+
+end Basic
+
+/-! ### Declarations about `single` -/
+
+
+section Single
+
+variable [Zero M] {a a' : α} {b : M}
+
+/-- `single a b` is the finitely supported function with value `b` at `a` and zero otherwise. -/
+def single (a : α) (b : M) : α →₀ M
+    where
+  support :=
+    haveI := Classical.decEq M
+    if b = 0 then ∅ else {a}
+  toFun :=
+    haveI := Classical.decEq α
+    Pi.single a b
+  mem_support_toFun a' := by
+    classical
```

Comment:
In lean 3, this call to classical results in all the lemmas within this section implicitly starting with `classical`